src/layers/trip-layer/trip-utils.js

```diff
@@ -0,0 +1,188 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import {Analyzer} from 'type-analyzer';
+
+import {
+  getSampleData,
+  timeToUnixMilli,
+  notNullorUndefined
+} from 'utils/data-utils';
+
+import {
+  parseGeoJsonRawFeature,
+  getGeojsonFeatureTypes
+} from 'layers/geojson-layer/geojson-utils';
+
+/**
+ * Parse geojson from string
+ * @param {array} geojson feature object values
+ * @returns {boolean} whether the geometry coordinates has length of 4
+ */
+export function coordHasLength4(samples) {
+  let hasLength4 = true;
+  for (let i = 0; i < samples.length; i += 1) {
+    hasLength4 = !samples[i].geometry.coordinates.find(c => c.length < 4);
+    if (!hasLength4) {
+      break;
+    }
+  }
+  return hasLength4;
+}
+
+/**
+ * Check whether geojson linestring's 4th coordinate is 1) not timestamp 2) unix time stamp 3) real date time
+ * @param {array} data array to be tested if its elements are timestamp
+ * @returns {string} the type of timestamp: unix/datetime/invalid(not timestamp)
+ */
+
+export function containValidTime(timestamps) {
+  const formattedTimeStamps = timestamps.map(ts => ({ts}));
+  const analyzedType = Analyzer.computeColMeta(formattedTimeStamps)[0];
+
+  if (!analyzedType || analyzedType.category !== 'TIME') {
+    return false;
+  }
+  return analyzedType;
+}
+
+/**
+ * Check if geojson features are trip layer animatable by meeting 3 conditions
+ * @param {array} features array of geojson feature objects
+ * @returns {boolean} whether it is trip layer animatable
+ */
+export function isTripGeoJsonField(allData = [], field) {
+  if (!allData.length) {
+    return false;
+  }
+  const getValue = d => d[field.tableFieldIndex - 1];
+  const maxCount = 10000;
+  const sampleRawFeatures =
+    allData.length > maxCount ? getSampleData(allData, maxCount, getValue) : allData.map(getValue);
+
+  const features = sampleRawFeatures.map(parseGeoJsonRawFeature);
+
+  let isTrip = false;
+  const featureTypes = getGeojsonFeatureTypes(features);
+
+  // condition 1: contain line string
+  const hasLineString = Boolean(featureTypes.line);
+  if (!hasLineString) {
+    return isTrip;
```

Comment:
rather than using isTrip variable, can't you just return false

Reply:
not sure if it's neccessary, this only used here